### PR TITLE
[Transaction]Delete transaction marker implement

### PR DESCRIPTION
**Pull request by congbobo184**
### Motivation
delete transaction marker for topic.

### Modifications

when cursor sign the markDeletePosition, we will check the next position is transaction marker. if it is we will delete the position.

### Verifying this change

add the test for it :
org.apache.bookkeeper.mledger.impl.ManagedCursorIndividualDeletedMessagesTest

### Does this pull request potentially affect one of the following parts:

*If `yes` was chosen, please highlight the changes*

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API: (no)
  - The schema: (no)
  - The default values of configurations: (no)
  - The wire protocol: (no)
  - The rest endpoints: (no)
  - The admin cli options: (no)
  - Anything that affects deployment: (no)

### Documentation

  - Does this pull request introduce a new feature? (yes)
  - If yes, how is the feature documented? (not applicable / docs / JavaDocs / not documented)
  - If a feature is not applicable for documentation, explain why?
  - If a feature is not documented yet in this PR, please create a followup issue for adding the documentation

### CODE REVIEW

File: pulsar-broker/src/test/java/org/apache/pulsar/broker/service/TransactionMarkerDeleteTest.java [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+    }
+
+    @AfterMethod
+    @Override
+    protected void cleanup() throws Exception {
+        super.internalCleanup();
+    }
+
+    @Test
+    public void TransactionMarkerDeleteTest() throws Exception {
```
Comment (codelipenghui) [inline pulsar-broker/src/test/java/org/apache/pulsar/broker/service/TransactionMarkerDeleteTest.java:53]:
```suggestion
    public void testTransactionMarkerDelete() throws Exception {
```

File: pulsar-broker/src/test/java/org/apache/pulsar/broker/service/TransactionMarkerDeleteTest.java [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+        Position position3 = managedLedger.addEntry("test".getBytes());
+        PersistentTopic topic = mock(PersistentTopic.class);
+        doReturn(managedLedger).when(topic).getManagedLedger();
+        managedLedger.openCursor("test");
+        PersistentSubscription persistentSubscription = new PersistentSubscription(topic, "test",
+                managedLedger.openCursor("test"), false);
+        persistentSubscription.acknowledgeMessage(Collections.singletonList(position1),
+                AckType.Individual, Collections.emptyMap());
+        assertEquals(0, ((PositionImpl) persistentSubscription.getCursor()
+                .getMarkDeletedPosition()).compareTo((PositionImpl) position3));
```
Comment (codelipenghui) [inline pulsar-broker/src/test/java/org/apache/pulsar/broker/service/TransactionMarkerDeleteTest.java:66]: I did not understand this test, looks not related to the changes that the PR introduced. At least you need to write one Transaction Marker into the managed ledger?

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentSubscription.java [hunk 3]
```diff
@@ -393,6 +398,27 @@ public void acknowledgeMessage(List<Position> positions, AckType ackType, Map<St
...
     }
 
+    private void deleteTransactionMarker(PositionImpl position, AckType ackType, Map<String,Long> properties) {
+        if (position != null) {
+            ManagedLedgerImpl managedLedger = ((ManagedLedgerImpl) cursor.getManagedLedger());
+            managedLedger.asyncReadEntry(position, new ReadEntryCallback() {
+                @Override
+                public void readEntryComplete(Entry entry, Object ctx) {
+                    MessageMetadata messageMetadata = Commands.parseMessageMetadata(entry.getDataBuffer());
+                    if (Markers.isTxnCommitMarker(messageMetadata)) {
```
Comment (codelipenghui) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentSubscription.java:408]: I think we should delete all transaction markers right? not only the commit marker.

**Review by codelipenghui — CHANGES_REQUESTED** [review]

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentSubscription.java [hunk 4]
```diff
@@ -368,6 +372,7 @@ public void acknowledgeMessage(List<Position> positions, AckType ackType, Map<St
         }
 
         if (!cursor.getMarkDeletedPosition().equals(previousMarkDeletePosition)) {
+            deleteTransactionMarker((PositionImpl) cursor.getMarkDeletedPosition(), ackType, properties);
```
Comment (codelipenghui) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentSubscription.java:375]: The mark delete position of the cursor may be modified, it's better to avoid to all `cursor.getMarkDeletedPosition()`  twice, it might return a different position. So you can use a local variable to maintain the currentMarkDeletePosition.

File: pulsar-broker/src/test/java/org/apache/pulsar/broker/service/TransactionMarkerDeleteTest.java [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+        managedLedger.addEntry(Markers
+                .newTxnCommitMarker(1, 1, 1, messageIdData).array());
+        Position position3 = managedLedger.addEntry(Markers
+                .newTxnCommitMarker(1, 1, 1, messageIdData).array());
+        managedLedger.openCursor("test");
+        persistentSubscription.acknowledgeMessage(Collections.singletonList(position1),
+                AckType.Individual, Collections.emptyMap());
+        Thread.sleep(1000L);
+        assertEquals(0, ((PositionImpl) persistentSubscription.getCursor()
+                .getMarkDeletedPosition()).compareTo((PositionImpl) position3));
```
Comment (codelipenghui) [inline pulsar-broker/src/test/java/org/apache/pulsar/broker/service/TransactionMarkerDeleteTest.java:75]: Please release the ledger and cursor resources after the test.

File: pulsar-broker/src/test/java/org/apache/pulsar/broker/service/TransactionMarkerDeleteTest.java [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+                managedLedger.openCursor("test"), false);
+        MessageIdData messageIdData = MessageIdData.newBuilder()
+                .setLedgerId(1)
+                .setEntryId(1)
+                .build();
+        Position position1 = managedLedger.addEntry("test".getBytes());
+        managedLedger.addEntry(Markers
+                .newTxnCommitMarker(1, 1, 1, messageIdData).array());
+        Position position3 = managedLedger.addEntry(Markers
+                .newTxnCommitMarker(1, 1, 1, messageIdData).array());
```
Comment (codelipenghui) [inline pulsar-broker/src/test/java/org/apache/pulsar/broker/service/TransactionMarkerDeleteTest.java:69]: // check backlog

**Review by codelipenghui — APPROVED** [review]

**Comment by lhotari** [comment]
@congbobo184 Since the Pulsar CI is very flaky at the moment, a lot of builds might fail because of flaky tests. There's more details on the #dev channel in Pulsar Slack.

**Comment by congbobo184** [comment]
@lhotari sorry, I'll pay attention later:).
